Add bookings

UserNameToHub opened the pull request.

UserNameToHub committed:
feat: добавить зависимость Postgresql
feat: добавить properties для application.properties
feat: добавить schema.sql
delete: ItemRepositoryImpl.java, Feedback.java, BookingTo.java

UserNameToHub committed:
refactor: изменить базовые классы BaseDtoMapper, BaseRepository, BaseService
delete: BaseDto.java

UserNameToHub committed:
feat: добавить Booking-entity
feat: добавить BookingRepository
feat: добавить BookingService, BookingServiceImpl
feat: добавить Dto
feat: добавить BookingController

UserNameToHub committed:
feat: добавить enumerations State, Status

UserNameToHub committed:
feat: добавить константы для времени и сортировки

UserNameToHub committed:
feat: добавить исключение DateException
feat: добавить DateException в ErrorHandler

UserNameToHub committed:
refactor: изменить метод findAll класса BookingServiceImpl

UserNameToHub committed:
feat: добавить класс UnavailableException

UserNameToHub committed:
feat: добавить entity User
style: изменить имя интерфейса BaseDtoMapper
delete: удалить статик метод isUnique, удалить класс UserRepositoryImpl.java
refactor: добавить метод findById класса UserService
refactor: изменить класс UserServiceImpl
feat: добаить enum UserStatus

UserNameToHub committed:
feat: добавить entity Item
refactor: изменить ItemController, ItemDtoMapper, ItemRepository, ItemService, ItemServiceImpl, ItemTo
feat: добавить ItemBookingGettingTo, ItemGettingDtoMapper, ItemGettingTo

UserNameToHub committed:
feat: добавить entity Comment
feat: добавить CommentRepository
feat: добавить comment dtoMapper
feat: добавить comment dto

UserNameToHub committed:
style: BaseDtoMapper, ItemDtoMapper, ItemRepository, ItemRequestDtoMapperMapper, ItemService, ItemTo

UserNameToHub committed:
style: BookingService, BookingServiceImpl, ItemRequestController

UserNameToHub committed:
feat: добавить настройки h2

UserNameToHub committed:
refactor: изменить ItemGettingDtoMapper

UserNameToHub committed:
style: добавить пробелы ItemGettingDtoMapper

EugeneLenkevich commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,78 @@
+    @Query("select new ru.practicum.shareit.item.dto.ItemBookingGettingTo(bk.id, bk.booker.id) " +
+            "from Booking as bk " +
+            "where bk.item.id = :id and bk.status = 'APPROVED' and bk.startDate < :dateTime " +
+            "order by bk.startDate desc")
```
У нас в параметрах метода есть Pageable, в нем есть возможность указать сортировку, тут, в запросе, тогда лучше сортировку убрать иначе получается такой смысловой конфликт :)

UserNameToHub replied:
Привет! Спасибо за коммент) 

Испаравил)

EugeneLenkevich commented on src/main/java/ru/practicum/shareit/item/dto/CommentGettingTo.java:
```diff
@@ -0,0 +1,15 @@
+
+@Data
+@Builder
+public class CommentGettingTo {
```
Не совсем понятно именование объекта, дословно "коментарий попадающий в"
"To" - это направление,  может тут подразумевалось существительное Dto (data transfer object) ?

В идеале хотчется чтобы код читался как обычный текст. Тут же, из именования объекта не совсем понятно про что он, может просто оставить CommentDto?  :)

И далее идет целая пачка dto, а надо ли столько, нельзя ли обойтись меньшим количеством, в этом же всем потом надо будет разбираться, поддерживать, эта специализация dto точно оправдана, нет ли тут переусложнения, не плодим ли лишние сущности? :)

UserNameToHub replied:
Еще раз спасибо!:)

Согласен, чет я Dtошек налепил) Постарался минисизировать)

EugeneLenkevich reviewed (changes requested):
Привет, Александр, спасибо за PR

Видно что задание хорошо проработано с технической стороны, ожинг только join fetch чего стоит)
Однако вот что хочется отметить. 
1. Если уж делаешь что-то то делай единообразно. В частности, сортировка в запросах делается тремя способами (order by, sort, pageable). Лучше выбрать только один способ и использовать только его, так будет единообразно, предлагаю использовать Pageable как самый универсальный
2. Вообще, чем код проще, тем он понятней и тем его потом легче, читай дешевле, поддерживать, может можно обойтись меньшим количеством dto, как это можно упростить :)

UserNameToHub committed:
refactor: изменить BookingController, BookingCreatingDtoMapper, BookingGettingDtoMapper, BookingRepository, BookingServiceImpl, BookingServiceImpl, CommentDtoMapper, ItemController, ItemDtoMapper, ItemDtoMapper, UserController, UserDtoMapper, UserService, UserServiceImpl

feat: добавить константу в Constants, ReflectionException

UserNameToHub committed:
style: изменить имя BookingCreatingDto, BookingGettingDto, CommentDto, ItemBookingGettingDto, ItemDto

delete: BookingItemDtoMapper, BookingItemGettingTo, BookingUserDtoMapper, BookingUserGettingTo, CommentCreatingTo, CommentCreatingTo, CommentUserTo, ItemGettingDtoMapper, ItemGettingTo

UserNameToHub committed:
delete: удалить дубликат зависимости
refactor: исправить выброс исключения

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо :)
